### ISSUE TITLE
about loss function

### ISSUE THREAD

**Issue by Broad-sky** [yjh0410/yolov2-yolov3_PyTorch#40]
"txty_loss_function = nn.BCEWithLogitsLoss(reduction='none')"  doesn't match the original paper?
can you explain it.

thanks!!

**Comment by developer0hye** [comment]
MSELoss for sigmoid output can make model drowning into local minima.
Many other yolo re-implementation projects use BCELoss or BCEWithLogitsLoss like this project.

**Comment by yjh0410** [comment]
Yeah，you're right. I tried BCE and MSE for obj loss, but I found that MSE always brought the better result. Therefore I select MSE not BCE.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "yjh0410/yolov2-yolov3_PyTorch"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2021年2月16日(星期二) 中午11:49
收件人:&nbsp;"yjh0410/yolov2-yolov3_PyTorch"<yolov2-yolov3_PyTorch@noreply.github.com&gt;;
抄送:&nbsp;"Subscribed"<subscribed@noreply.github.com&gt;;
主题:&nbsp;Re: [yjh0410/yolov2-yolov3_PyTorch] about loss function (#40)





 
MSELoss for sigmoid output can make model drowning into local minima.
 Many other yolo re-implementation projects use BCELoss or BCEWithLogitsLoss like this project.

**Comment by developer0hye** [comment]
@yjh0410 

Hi! I recently developed and distributed python package [plotbbox](https://github.com/developer0hye/plotbbox) that is the tool to plot pretty bounding box. I share this with you!

**Comment by yjh0410** [comment]
Looks so pretty! Thanks a lot !

**Comment by developer0hye** [comment]
@yjh0410 Thanks a lot! what object detection algorithm do you study now? Do you have recommendations?

**Comment by yjh0410** [comment]
Sorry, I haven't deeply studied object detection for a long time, just optimize existing projects. My tutor is considering letting me to do research on Temporal Action Detection. 

Last week, I read the paper of **OneNet** which removes the post-process including NMS to make the object detection pipeline more concise. The purpose of **OneNet** is to make one prediction correspond to one goal, so it is unnecessary to deploy NMS to filter bboxes. Maybe, I think that how to make sure that one prediction corresponds to one object rather than multi predictions  corresponds to one object (which means we have to use NMS.) is a good research direction.

**Comment by developer0hye** [comment]
@yjh0410 
Thanks! Their Idea looks so simple in terms of implementation, but effect is great!

**Comment by Broad-sky** [comment]
@yjh0410 

cls_loss = torch.sum(cls_loss_function(pred_cls, gt_cls) * gt_mask) / batch_size

Why divide by batch_size？？

Could you explain that！
Hope to reply，Thanks!

**Comment by developer0hye** [comment]
@Broad-sky 
to normalize the scale of loss by batch size

**Comment by yjh0410** [comment]
@Broad-sky 
It is known to us that we must normalize the loss by batch size. 
I also try to divide it by all samples, but it doesn't work.

**Comment by developer0hye** [comment]
@Broad-sky 

We only calculate the loss function for positive samples. If we dived its the output of the loss function by the number of all samples, the class loss will be very small value.  So, Its gradient value will be close to 0.
